tools/src/wyvern/target/corewyvernIL/decl/DefDeclaration.java

```diff
@@ -79,6 +88,16 @@ public DeclType typeCheck(TypeContext ctx, TypeContext thisCtx) {
 		for (FormalArg arg : formalArgs) {
 			methodCtx = methodCtx.extend(arg.getName(), arg.getType());
 		}
+
+		for (String freeVar : this.getFreeVariables()) {
+			ValueType t = (new Variable(freeVar)).typeCheck(methodCtx);
+			if (t instanceof StructuralType) {
+				this.resourceFlag = ((StructuralType) t).isResource();
```

Comment:
This is buggy.  You might assign resourceFlag = true the first time through the loop, and resourceFlag=false the second time.  Really you want to take the logical or of the isResource() of all captured variables.